[FLINK-15782] [connectors/jdbc] JDBC sink DataStream API

rkhachatryan opened the pull request:
## What is the purpose of the change

1. expose existing sinks at the DataStream API level (for consistency with the planned exactly once; and to use on its own); existing API shouldn’t change (mostly table API)
2. refactor current code to allow re-use (in the planned exactly once JDBC sink); 
some issues are
- it operates on `Row` objects and (for upsert) 
- expects a tuple of `<boolean, Row>`
- generation of SQL doesn't make sense for POJO, because parameters must be set explicitly
- upsert emulation (`select` + `update` OR `isnert`) doesn't make sense for DataStream API

## Out of scope
- Table API
- existing implementation classes not required to implement the new API

## Brief change log
  - *extract connection management from `OutputFormat` to new `JdbcConnectionProvider`*
  - *replace `Row` with type parameter; move Table API specific code (like Tuple2<Boolean, T>) to the appropriate classes; as a result*:
   -- `GenericJdbcSinkFunction` (from `JDBCUpsertSinkFunction`)
   -- `JdbcBatchingOutputFormat` (from `JDBCUpsertOutputFormat`)
  - *add DataStream API (facade class)*
  - *add docs*
  - *minor changes: (change case `JDBC` -> `Jdbc`, parameter objects)*
  - *refactor tests*

## Verifying this change

This change is already covered by existing tests, such as *JDBCFullTest, JDBCOutputFormatTest, JdbcTableOutputFormatTest*.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: yes
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): yes
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? yes
  - If yes, how is the feature documented? docs

aljoscha commented on docs/redirects/jdbc.md:
```diff
@@ -0,0 +1,24 @@
+---
```
Why are these redirects added?

rkhachatryan replied:
Just for uniformity with other pages.

aljoscha replied:
The redirects are only there because previously the documentation structure was different. For new pages we don't need redirects.

aljoscha commented on flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/JdbcStatementBuilder.java:
```diff
@@ -0,0 +1,46 @@
+	 * Creates a {@link JdbcStatementBuilder} for {@link Row} using the provided SQL types array.
+	 * Uses {@link org.apache.flink.api.java.io.jdbc.JDBCUtils#setRecordToStatement}
+	 */
+	static JdbcStatementBuilder<Row> forRow(int[] types) {
```
I think in the long run we don't want to have `Row` in non-Table packages. We therefore shouldn't further expose it in our public APIs.

kl0u commented on docs/dev/connectors/jdbc.md:
```diff
@@ -0,0 +1,97 @@
+
+This connector provides a sink that writes data to a JDBC database.
+
+To use this it, add the following dependency to your project (along with your JDBC-driver):
```
Remove the "this".

kl0u commented on flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/JdbcBatchingOutputFormat.java:
```diff
@@ -0,0 +1,298 @@
+			RecordExtractor<In, JdbcIn> recordExtractor) {
+		super(connectionProvider);
+		this.executionOptions = executionOptions;
+		this.statementRunnerCreator = statementExecutorCreator;
```
Maybe uniformise the names, either use `runner` or `executor` for both the class (type) and the variable (instance). It is just easier to follow the code I think.

kl0u commented on flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/JdbcBatchingOutputFormat.java:
```diff
@@ -0,0 +1,298 @@
+	@Override
+	public void open(int taskNumber, int numTasks) throws IOException {
+		super.open(taskNumber, numTasks);
+		jdbcStatementExecutor = statementRunnerCreator.apply(getRuntimeContext());
```
Maybe introduce a method like: 

```
private JdbcExec initStatementExecutor() throws IOException {
		final JdbcExec exec = statementRunnerCreator.apply(getRuntimeContext());
		try {
			jdbcStatementExecutor.open(connection);
		} catch (SQLException e) {
			throw new IOException("unable to open JDBC writer", e);
		}
		return exec;
	}
```

To offload a bit of the code in the `open()`. What do you think?

rkhachatryan replied:
Yes, I think that's a good idea.

kl0u commented on flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/JdbcBatchingOutputFormat.java:
```diff
@@ -0,0 +1,298 @@
+			this.scheduler = Executors.newScheduledThreadPool(1, new ExecutorThreadFactory("jdbc-upsert-output-format"));
+			this.scheduledFuture = this.scheduler.scheduleWithFixedDelay(() -> {
+				synchronized (JdbcBatchingOutputFormat.this) {
+					if (!closed) {
```
Are we sure about the concurrency here? Here `closed` is guarded by a lock but all the remaining usages are not. Here we are simply reading the value which is `volatile`, but in the `close()` we have a `if (!closed) { closed = true;  ... ` which is not atomic and then we `cancel` the `scheduledFuture` executing this code.

rkhachatryan replied:
`close()` is also `synchronized` so these sections can't execute simultaneously.

`volatile` seems unnecessary here; I didn't remove this modifier because I tried to keep the internal logic untouched where possible (performance-wise it doesn't matter since it's not in the hot path).

kl0u commented on flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/JdbcBatchingOutputFormat.java:
```diff
@@ -0,0 +1,298 @@
+
+		for (int i = 1; i <= executionOptions.getMaxRetries(); i++) {
+			try {
+				attemptFlush();
```
Here shouldn't we restart the timer? It seems like we may be in the situation where we flushed this batch because we reached the max size, so for the new one we have to set a new timer (renew the `BatchIntervalMs`), right?

rkhachatryan replied:
Yes. I would also replace `Thread.sleep(1000 * i);` with just rescheduling (retaining blocking semantics of `write`).
But I'm not sure if such changes should be in this PR.
WDYT?

kl0u replied:
I agree that this seems to be more of a bug that deserves its own PR. Would you like to open a JIRA for this and assign it to you?

rkhachatryan replied:
Yes, that's a good idea.
Created a ticket for that: https://issues.apache.org/jira/browse/FLINK-16328

kl0u commented on flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/dialect/JDBCDialects.java:
```diff
@@ -152,7 +152,7 @@ public String quoteIdentifier(String identifier) {
 		}
 
-		@Override
+@Override
```
Probably a typo?

kl0u commented on flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/executor/KeyedBatchStatementExecutor.java:
```diff
@@ -0,0 +1,79 @@
+import java.util.HashSet;
+import java.util.Set;
+import java.util.function.Function;
+
```
Here a javadoc for the types and the functionality would help. I noticed that none of the classes have one, so it may be not part of this PR, but it would be a helpful addition I think.

wuchong commented on flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/JdbcExecutionOptions.java:
```diff
@@ -0,0 +1,92 @@
+ * JDBC sink batch options.
+ */
+@PublicEvolving
+public class JdbcExecutionOptions implements Serializable {
```
It seems that this option is only used for batch writing. `Execution` sounds like it also contains reading/query, because `PreparedStatement` also exposes a `executeQuery` method. 
Shall we rename it to `JdbcWritingOptions` or `JdbcSinkOptions`?

rkhachatryan replied:
`JdbcWritingOptions` doesn't seem to match `delete` queries.
And `JdbcSinkOptions` sounds too generic to me. I'd expect something like JDBC URL inside.

As for `executeQuery`, I don't see why `JdbcExecutionOptions` can't be used for it (batch could mean `limit` then).

wuchong replied:
I don't think batch could mean `limit`. From my understanding, `JdbcExecutionOptions` is just a tuning options which should affect the final result. And the `batchIntervalMs` doesn't make sense for readers. 
But if we dont' have a better name for it, I'm also fine with `JdbcExecutionOptions`.

rkhachatryan replied:
I agree, `JdbcExecutionOptions` isn't ideal, but other names aren't better.

wuchong commented on flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/JdbcSink.java:
```diff
@@ -0,0 +1,63 @@
+ * Facade to create JDBC {@link SinkFunction sinks}.
+ */
+@PublicEvolving
+public class JdbcSink {
```
Shall we expose a corresponding `JdbcSource`, or it will be introduced later?

rkhachatryan replied:
This PR only targets JDBC sinks.
For sources I think a similar refactoring would be required.

rkhachatryan commented on flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/JdbcConnectionOptions.java:
```diff
@@ -0,0 +1,94 @@
+	public String getPassword() {
+		return password;
+	}
+
```
This name clashes with `JDBCOptions#builder` which is part of current public API.
I didn't want to have similar methods named differently so I decided just to create builder with `new`.

rkhachatryan replied:
This name clashes with `JDBCOptions#builder` which is part of current public API.
I didn't want to have similar methods named differently so I decided just to create builder with `new`.

rkhachatryan commented on flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/executor/JdbcBatchStatementExecutor.java:
```diff
@@ -0,0 +1,102 @@
+	 */
+	void open(Connection connection) throws SQLException;
+
+	void process(T record) throws SQLException;
```
Yes, I think it's a good idea.

rkhachatryan commented on flink-connectors/flink-jdbc/src/test/java/org/apache/flink/api/java/io/jdbc/JDBCTestCheckpoint.java:
```diff
@@ -0,0 +1,35 @@
+/**
+ * Holds id and indices of items in {@link JdbcTestFixture#TEST_DATA}.
+ */
+public class JDBCTestCheckpoint {
```
Yes, this is a leftover after splitting the work.
Removed it, thanks.

rkhachatryan commented on flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/executor/KeyedBatchStatementExecutor.java:
```diff
@@ -0,0 +1,79 @@
+	public void executeBatch() throws SQLException {
+		if (!batch.isEmpty()) {
+			for (K entry : batch) {
+				parameterSetter.accept(st, entry);
```
You are right, thanks for pointing this out.
I'll change it to use `SimpleBatchStatementExecutor`.

wuchong commented on flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/JdbcBatchingOutputFormat.java:
```diff
@@ -46,15 +47,15 @@
 	}
 
-	interface ExecutorCreator<T extends JdbcBatchStatementExecutor<?>> extends Function<RuntimeContext, T>, Serializable {
+	interface StatementExecutorCreator<T extends JdbcBatchStatementExecutor<?>> extends Function<RuntimeContext, T>, Serializable {
```
What about `StatementExecutorFactory

rkhachatryan commented:
Thanks a lot for reviewing @aljoscha  ,  @kl0u ,   @wuchong .
I've addressed the issues and currently waiting for another PR to pass CI (didn't push yet).

rkhachatryan commented:
Please review the changes (I've added them as a separate commit).

rkhachatryan commented:
I've rebased the PR onto the latest master.
@kl0u , @wuchong, could you please take a look?

kl0u commented on flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/executor/InsertOrUpdateJdbcExecutor.java:
```diff
@@ -0,0 +1,124 @@
+	private transient Map<K, V> batch = new HashMap<>();
+
+	public InsertOrUpdateJdbcExecutor(
+		String existSQL,
```
What about one more tab on the arguments so that we can distinguish between the args and the body of the method?

rkhachatryan replied:
How do I always mess with indentation 🤦‍♂️

kl0u commented on flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/executor/InsertOrUpdateJdbcExecutor.java:
```diff
@@ -0,0 +1,124 @@
+		JdbcStatementBuilder<V> updateSetter,
+		Function<R, K> keyExtractor,
+		Function<R, V> valueExtractor) {
+		this.existSQL = existSQL;
```
I think it makes sense to add null checks here so that the contracts of the class are explicit.

kl0u commented on flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/GenericJdbcSinkFunction.java:
```diff
@@ -19,18 +19,18 @@
+	private final AbstractJdbcOutputFormat<T > outputFormat;
+
+	GenericJdbcSinkFunction(AbstractJdbcOutputFormat<T > outputFormat) {
 		this.outputFormat = outputFormat;
```
Add a null check?

kl0u commented on flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/JdbcBatchingOutputFormat.java:
```diff
@@ -0,0 +1,316 @@
+			StatementExecutorFactory<JdbcExec> statementExecutorFactory,
+			RecordExtractor<In, JdbcIn> recordExtractor) {
+		super(connectionProvider);
+		this.executionOptions = executionOptions;
```
I would also add null checks here wherever needed.

kl0u commented on flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/JdbcBatchingOutputFormat.java:
```diff
@@ -0,0 +1,316 @@
+
+	private final JdbcExecutionOptions executionOptions;
+	private final StatementExecutorFactory<JdbcExec> statementExecutorFactory;
+	final RecordExtractor<In, JdbcIn> jdbcRecordExtractor;
```
What about making it private and adding a package private or protected getter. This is a question of personal taste though so feel free to ignore if you disagree :)

rkhachatryan replied:
Hmm...I think we can have a different "template method" - `addToBatch` instead of `doWriteRecord`; so this field is accessed only in the base class and can be `private`:
```
@Override
void addToBatch(Tuple2<Boolean, Row> original, Row extracted) throws SQLException {
...
```

kl0u commented on flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/JdbcDmlOptions.java:
```diff
@@ -0,0 +1,127 @@
+	}
+
+	/**
+	 * JDBCUpsertOptionsBuilder.
```
Probably add sth like "A builder for..." I understand that this is kind of redundant but at least have sth more meaningful :P

rkhachatryan replied:
OK :)

kl0u commented on flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/JdbcDmlOptions.java:
```diff
@@ -0,0 +1,127 @@
+		private String[] fieldNames;
+		private String[] keyFields;
+		private JDBCDialect dialect;
+		private JDBCDialect customDialect;
```
This seems to not be used anywhere.

rkhachatryan replied:
Leftover from some previous iteration. 
Removed, thanks.

kl0u commented on flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/executor/JdbcBatchStatementExecutor.java:
```diff
@@ -16,29 +16,27 @@
  */
-public interface JDBCWriter extends Serializable {
+@Internal
+public interface JdbcBatchStatementExecutor<T> {
```
Here all the methods are expected to only throw `SQLException` and then, in many cases, we wrap the exceptions in other types. Does it make sense to simply have these methods throw an `Exception`?

rkhachatryan replied:
On higher levels, it is wrapped because there the allowed exception is `IOException`, not `SQLException`. So we can only declare `throws IOException` without wrapping.

But IMO `SQLException` is better here because:
- caller can actually do something with it
- it has an appropriate abstraction level

rkhachatryan replied:
After further looking at code, I noticed it was wrapped in one place into `RuntimeException` instead of `IOException` so I fixed wrapping.
